bivac.1
.TH bivac 1 "2 May 2018"
.SH NAME
bivac \- Docker volumes backup
.SH SYNOPSIS
\fBbivac\fP [OPTIONS]
.SH DESCRIPTION
Bivac lets you backup all your names Docker volumes using Duplicity or RClone.

Bivac supports multiple engines for performing the backup:

* Duplicity (default engine)

* RClone: use for heavy data that Duplicity cannot manage efficiently

* Restic

.SH OPTIONS
.SS Application Options
.TP
\fB\fB\-V\fR, \fB\-\-version\fR\fP
Display version.
.TP
\fB\fB\-l\fR, \fB\-\-loglevel\fR <default: \fI"info"\fR>\fP
Set loglevel ('debug', 'info', 'warn', 'error', 'fatal', 'panic').
.TP
\fB\fB\-b\fR, \fB\-\-blacklist\fR <default: \fI$BIVAC_VOLUMES_BLACKLIST\fR>\fP
Volumes to blacklist in backups.
.TP
\fB\fB\-m\fR, \fB\-\-manpage\fR\fP
Output manpage.
.TP
\fB\fB\-\-no-verify\fR <default: \fI$BIVAC_NO_VERIFY\fR>\fP
Do not verify backup.
.TP
\fB\fB\-j\fR, \fB\-\-json\fR <default: \fI$BIVAC_JSON_OUTPUT\fR>\fP
Log as JSON (to stderr).
.TP
\fB\fB\-E\fR, \fB\-\-engine\fR <default: \fI"duplicity"\fR>\fP
Backup engine to use.
.TP
\fB\fB\-o\fR, \fB\-\-orchestrator\fR <default: \fI"docker"\fR>\fP
Container orchestrator to use.
.TP
\fB\fB\-u\fR, \fB\-\-target-url\fR <default: \fI$BIVAC_TARGET_URL\fR>\fP
The target URL to push to.
.TP
\fB\fB\-H\fR, \fB\-\-hostname-from-rancher\fR <default: \fI$BIVAC_HOSTNAME_FROM_RANCHER\fR>\fP
Retrieve hostname from Rancher metadata.
.TP
\fB\fB\-\-check-every\fR <default: \fI"24h"\fR>\fP
Time between backup checks.
.TP
\fB\fB\-\-remove-older-than\fR <default: \fI"30D"\fR>\fP
Remove backups older than the specified interval.
.TP
\fB\fB\-\-label-prefix\fR <default: \fI$BIVAC_LABEL_PREFIX\fR>\fP
The volume prefix label.
.SS Duplicity Options
.TP
\fB\fB\-\-duplicity-image\fR <default: \fI"camptocamp/duplicity:latest"\fR>\fP
The duplicity docker image.
.TP
\fB\fB\-\-full-if-older-than\fR <default: \fI"15D"\fR>\fP
The number of days after which a full backup must be performed.
.SS RClone Options
.TP
\fB\fB\-\-rclone-image\fR <default: \fI"camptocamp/rclone:1.33-1"\fR>\fP
The rclone docker image.
.SS Restic Options
.TP
\fB\fB\-\-restic-image\fR <default: \fI"restic/restic:latest"\fR>\fP
The restic docker image.
.TP
\fB\fB\-\-restic-password\fR <default: \fI$RESTIC_PASSWORD\fR>\fP
The restic backup password.
.SS Metrics Options
.TP
\fB\fB\-g\fR, \fB\-\-gateway-url\fR <default: \fI$PUSHGATEWAY_URL\fR>\fP
The prometheus push gateway URL to use.
.SS AWS Options
.TP
\fB\fB\-\-aws-access-key-id\fR <default: \fI$AWS_ACCESS_KEY_ID\fR>\fP
The AWS access key ID.
.TP
\fB\fB\-\-aws-secret-key-id\fR <default: \fI$AWS_SECRET_ACCESS_KEY\fR>\fP
The AWS secret access key.
.SS Swift Options
.TP
\fB\fB\-\-swift-username\fR <default: \fI$SWIFT_USERNAME\fR>\fP
The Swift user name.
.TP
\fB\fB\-\-swift-password\fR <default: \fI$SWIFT_PASSWORD\fR>\fP
The Swift password.
.TP
\fB\fB\-\-swift-auth_url\fR <default: \fI$SWIFT_AUTHURL\fR>\fP
The Swift auth URL.
.TP
\fB\fB\-\-swift-tenant-name\fR <default: \fI$SWIFT_TENANTNAME\fR>\fP
The Swift tenant name.
.TP
\fB\fB\-\-swift-region-name\fR <default: \fI$SWIFT_REGIONNAME\fR>\fP
The Swift region name.
.SS Docker Options
.TP
\fB\fB\-e\fR, \fB\-\-docker-endpoint\fR <default: \fI"unix:///var/run/docker.sock"\fR>\fP
The Docker endpoint.
.SS Kubernetes Options
.TP
\fB\fB\-\-k8s-namespace\fR <default: \fI$K8S_NAMESPACE\fR>\fP
Namespace where you want to run Bivac.
.TP
\fB\fB\-\-k8s-kubeconfig\fR <default: \fI$K8S_KUBECONFIG\fR>\fP
Path to your kubeconfig file.
.SS Help Options
.TP
\fB\fB\-h\fR, \fB\-\-help\fR\fP
Show this help message


CHANGELOG.md
# [0.26.1](https://github.com/camptocamp/bivac/releases/tag/0.26.1) (2017-05-26)

* Rclone:

  - Pin camptocamp/rclone:1.33-1 as newer images changes environment variable API


# [0.26.0](https://github.com/camptocamp/bivac/releases/tag/0.26.0) (2017-01-03)

* Metrics:

  - Use volume as primary key in metrics, avoids getting all metrics (PR #117, fix #115)

# [0.25.6](https://github.com/camptocamp/bivac/releases/tag/0.25.6) (2016-10-17)

* Bugfix:

  - Fix lint issue

# [0.25.5](https://github.com/camptocamp/bivac/releases/tag/0.25.5) (2016-10-17)

* Features:

  - Retry on errors (fix #112)

# [0.25.4](https://github.com/camptocamp/bivac/releases/tag/0.25.4) (2016-10-17)

* Bugfix:

  - Do not crash when GetMetrics fails (fix #108)

# [0.25.3](https://github.com/camptocamp/bivac/releases/tag/0.25.3) (2016-09-26)

* Dependencies:

  - Switch to docker/docker since docker/engine-api is deprecated

* Bugfix:

  - Override p.SetVolumeBackupDir() for all providers (fix #107, really fix #103)

# [0.25.2](https://github.com/camptocamp/bivac/releases/tag/0.25.2) (2016-09-05)

* Build:

  - Build with Go 1.7 in Travis CI

* Bugfix:

  - Check exit code in PrepareBackup() (fix #105)

# [0.25.1](https://github.com/camptocamp/bivac/releases/tag/0.25.1) (2016-08-31)

* Bugfix:

  - Drop databases before recreating (fix #104)
  - Call p.GetBackupDir(), not p.backupDir (fix #103)

# [0.25.0](https://github.com/camptocamp/bivac/releases/tag/0.25.0) (2016-08-22)

## Warning:

We don't use a specific path separator when using the duplicity engine for swift anymore, as the bug that prevented to backup to a pseudo-folder is fixed in duplicity 0.7.08.

# [0.24.5](https://github.com/camptocamp/bivac/releases/tag/0.24.5) (2016-08-18)

* Bugfix:

  - Really take the last chainEndTime element (fix #96)

# [0.24.4](https://github.com/camptocamp/bivac/releases/tag/0.24.4) (2016-08-17)

* Bugfix:

  - Fix c.ImageInspectWithRaw() call for new version of docker/engine-api

# [0.24.3](https://github.com/camptocamp/bivac/releases/tag/0.24.3) (2016-08-17)

* Bugfix:

  - Take the last chainEndTime element (fix #96)

* Tests:

  - Fix some tests

* Quality:

  - Add go report badge

# [0.24.2](https://github.com/camptocamp/bivac/releases/tag/0.24.2) (2016-08-10)

* Bugfix:

  - Fix UpdateEvent() not updating event

* Logging:

  - Make debug less verbose

* Tests:

  - Add tests

# [0.24.1](https://github.com/camptocamp/bivac/releases/tag/0.24.1) (2016-08-10)

* Bugfix:

  - Fix wrong condition checking on Event#Equals()

* Tests:

  - Add tests for Event#Equals()

# [0.24.0](https://github.com/camptocamp/bivac/releases/tag/0.24.0) (2016-08-10)

* Internals:

  - Linting
  - Avoid asserting config params as strings (fix #89)
  - Use a sub-struct for sub-config in Volumes (fix #90)
  - Use net/url for targetURL, fix RClone URLs (fix #92)

* Configuration:

  - Add more volume configuration (fix #91)
  - Use a global targetURL (fix #93)

* Tests:

  - Fix volume tests
  - Fix util tests

* Documentation:

  - Update README
  - Improve and update manpage


# [0.23.0](https://github.com/camptocamp/bivac/releases/tag/0.23.0) (2016-08-09)

* Internals:

  - Use reflect to compute Volume config (fix #87)

* Configuration:

  - Add a .bivac.overrides file as a replacement for Docker labels (fix #87)

* Metrics:

  - Add backupStartTime and backupEndTime metrics for volumes

# [0.22.0](https://github.com/camptocamp/bivac/releases/tag/0.22.0) (2016-08-08)

* Internals:

  - Fix volume labels fetching (fix #83)
  - Refactor metrics, using Metric and Event structs

* Metrics:

  - Change metric names according to Prometheus docs
  - Fetch existing metrics from Prometheus gateway (fix #84)

* Testing:

  - Add metrics to coverage

* Documentation:

  - Add return codes to README
  - Document engines in README

# [0.21.0](https://github.com/camptocamp/bivac/releases/tag/0.21.0) (2016-07-28)

* Internals:

  - Rename lib/ as handler/
  - Add NewBivac() method
  - Don't use panic
  - Inspect volumes in GetVolumes()
  - Propagate errors instead of crashing (fix #65)
  - Add metrics.go

# [0.20.0](https://github.com/camptocamp/bivac/releases/tag/0.20.0) (2016-07-28)

* Features:

  - Output version in logs
  - Make backup engine pluggable
  - Add RClone engine (fix #71, #76)
  - Add option to get host name from Rancher metadata (fix #73)

* Internals:

  - Check errors in PushToPrometheus
  - Read response from Prometheus

# [0.13.3](https://github.com/camptocamp/bivac/releases/tag/0.13.3) (2016-07-05)

* Bugfix:

  - Blacklist lost+found (fix #68)
  - Wait for container to exit before retrieving logs (fix #69)

# [0.13.2](https://github.com/camptocamp/bivac/releases/tag/0.13.2) (2016-07-05)

* Bugfix:

  - Fix logging in exec create
  - Create new Docker client for PrepareBackup (fix #70)

* Testing:

  - Add docker image testing

# [0.13.1](https://github.com/camptocamp/bivac/releases/tag/0.13.1) (2016-06-30)

* Bugfix:

  - Fix image pulling (fix #61)
  - Support "none" as value for last full backup date
    (avoid crashing on first backup)

* Internals:

  - Change some panic errors into fatal to avoid traces

* Testing:

  - Add more tests

# [0.13.0](https://github.com/camptocamp/bivac/releases/tag/0.13.0) (2016-06-29)

* Breaking changes:

  - JSON_OUTPUT renamed to BIVAC_JSON_OUTPUT

* Internals:

  - Reorganize packages in lib/ (fix #60)

* Bugfix:

  - Do not intent to remove links in removeContainer()

* Testing:

  - Test with verbose
  - Add more tests
  - Add coverage

# [0.12.0](https://github.com/camptocamp/bivac/releases/tag/0.12.0) (2016-06-28)

* Features:

  - Log as JSON with -j (fix #56)
  - Use logs with fields (fix #57)
  - Add stdout in debug output

* Internals:
  - Rewrite using github.com/docker/engine-api (fix #58)

* Bugfix:

  - Fix stdout redirection bug (fix #51)

# [0.11.0](https://github.com/camptocamp/bivac/releases/tag/0.11.0) (2016-06-28)

* Features:


bivac.go
package main

import (
	"fmt"
	"os"

	log "github.com/Sirupsen/logrus"
	"github.com/camptocamp/bivac/engines"
	"github.com/camptocamp/bivac/handler"
	"github.com/camptocamp/bivac/orchestrators"
	"github.com/camptocamp/bivac/providers"
	"github.com/camptocamp/bivac/util"
	"github.com/camptocamp/bivac/volume"
)

var version = "undefined"

func main() {
	var err error
	var exitCode int

	c, err := handler.NewBivac(version)
	util.CheckErr(err, "Failed to setup Bivac handler: %v", "fatal")

	log.Infof("Bivac v%s starting backup...", version)

	orch, err := orchestrators.GetOrchestrator(c)
	if err != nil {
		log.Fatalf("Failed to get an orchestrator: %s", err)
	}

	vols, err := orch.GetVolumes()
	util.CheckErr(err, "Failed to get Docker volumes: %v", "fatal")

	for _, vol := range vols {
		vol.LogTime("backupStartTime")
		err = backupVolume(orch, vol)
		vol.LogTime("backupEndTime")
		if err != nil {
			log.Errorf("Failed to backup volume %s: %v", vol.Name, err)
			exitCode = 1
			continue
		}
	}

	log.Infof("End backup...")
	os.Exit(exitCode)
}

func backupVolume(o orchestrators.Orchestrator, vol *volume.Volume) (err error) {
	p := providers.GetProvider(o, vol)
	log.WithFields(log.Fields{
		"volume":   vol.Name,
		"provider": p.GetName(),
	}).Info("Found data provider")

	vp, err := providers.PrepareBackup(p)
	if err != nil {
		err = fmt.Errorf("failed to prepare backup: %v", err)
		return
	}

	var e engines.Engine
	if vp != nil {
		e = engines.GetEngine(o, vp)
	} else {
		e = engines.GetEngine(o, vol)
	}

	log.WithFields(log.Fields{
		"volume": vol.Name,
		"engine": e.GetName(),
	}).Info("Found backup engine")

	err = e.Backup()
	if err != nil {
		err = fmt.Errorf("failed to backup volume: %v", err)
		return
	}
	return
}
